rechunker/api.py

```diff
@@ -10,12 +10,100 @@
 from rechunker.algorithm import rechunking_plan
 
 
-def rechunk_zarr2zarr_w_dask(
+def _shape_dict_to_tuple(dims, shape_dict):
+    # convert a dict of shape
+    shape = [shape_dict[dim] for dim in dims]
+    return tuple(shape)
+
+
+def _get_dims_from_zarr_array(z_array):
+    # use Xarray convention
+    # http://xarray.pydata.org/en/stable/internals.html#zarr-encoding-specification
+    return z_array.attrs["_ARRAY_DIMENSIONS"]
+
+
+def _zarr_empty(shape, store_or_group, chunks, dtype, name=None):
+    # wrapper that maybe creates the array within a group
+    if name is not None:
+        assert isinstance(store_or_group, zarr.hierarchy.Group)
+        return store_or_group.empty(name, shape=shape, chunks=chunks, dtype=dtype)
+    else:
+        return zarr.empty(shape, chunks=chunks, dtype=dtype, store=store_or_group)
+
+
+def rechunk(
+    source, target_chunks, max_mem, target_store, temp_store=None,
+):
+    """
+    Rechunk a Zarr Array or Group
+
+    Parameters
+    ----------
+    source : zarr.Array or zarr.Group
+    target_chunks : tuple or dict
+        The desired chunks of the array after rechunking.
+    max_mem : int
+        The amount of memory (in bytes) that workers are allowed to use
+    target_store : str, MutableMapping, or zarr.Store object
+        The location in which to store the final, rechunked result
+    temp_store : str, MutableMapping, or zarr.Store object, optional
+        Location of temporary store for intermediate data. Can be deleted
+        once rechunking is complete.
+    """
+
+    # these options are not tested yet; don't include in public API
+    kwargs = dict(
+        source_storage_options={}, temp_storage_options={}, target_storage_options={},
+    )
+
+    if isinstance(source, zarr.hierarchy.Group):
```

Comment:
This looks like a good use for singledispatch: https://docs.python.org/3/library/functools.html#functools.singledispatch


```python
@singledispatch
def rechunk_object(...)
    raise ValueError

@rechunk_object.register
def _(source: zarr.hierarch.Group, ...)
    ...
```